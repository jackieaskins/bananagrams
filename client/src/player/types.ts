export type Player = {
  userId: string;
  username: string;
  isReady: boolean;
  isTopBanana: boolean;
  isAdmin: boolean;
  gamesWon: number;
};
