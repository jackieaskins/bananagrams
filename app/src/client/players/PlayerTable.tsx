import {
  Table,
  TableCaption,
  TableContainer,
  Tbody,
  Th,
  Thead,
  Tr,
} from "@chakra-ui/react";
import { useGame } from "../games/GameContext";
import { socket } from "../socket";
import PlayerTableRow from "./PlayerTableRow";

const MAX_PLAYERS = 16;

export default function PlayerTable(): JSX.Element {
  const { gameInfo } = useGame();

  const players = gameInfo.players;
  const isCurrentPlayerAdmin =
    players.find(({ userId }) => userId === socket.id)?.isAdmin ?? false;

  return (
    <TableContainer>
      <Table variant="simple">
        <TableCaption>
          Current player count: {players.length}/{MAX_PLAYERS} max
        </TableCaption>

        <Thead>
          <Tr>
            <Th textAlign="right">Status</Th>
            <Th>Player</Th>
            <Th isNumeric>Games won</Th>
            {isCurrentPlayerAdmin && players.length > 1 && <Th />}
          </Tr>
        </Thead>

        <Tbody>
          {players.map((player) => (
            <PlayerTableRow
              key={player.userId}
              isCurrentPlayerAdmin={isCurrentPlayerAdmin}
              player={player}
              playerCount={players.length}
            />
          ))}
        </Tbody>
      </Table>
    </TableContainer>
  );
}
