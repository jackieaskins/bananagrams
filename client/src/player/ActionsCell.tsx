import { DeleteOutlined } from '@ant-design/icons';
import { Button } from 'antd';

import { useCurrentPlayer } from '../game/stateHooks';
import { kickPlayer } from '../socket';
import { Player } from './types';

type ActionsCellProps = {
  player: Player;
};

const ActionsCell = ({
  player: { userId },
}: ActionsCellProps): JSX.Element | null => {
  const currentPlayer = useCurrentPlayer();

  if (!currentPlayer) {
    return null;
  }

  const { isAdmin, userId: currentPlayerId } = currentPlayer;

  if (isAdmin && userId !== currentPlayerId) {
    return (
      <Button
        icon={<DeleteOutlined />}
        onClick={() => {
          kickPlayer({ userId });
        }}
        type="text"
      />
    );
  }

  return null;
};

export default ActionsCell;
