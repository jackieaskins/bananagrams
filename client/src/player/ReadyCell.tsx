import { CheckOutlined, CloseOutlined } from '@ant-design/icons';
import { Checkbox } from 'antd';

import { getUserId, setReady } from '../socket';
import { Player } from './types';

type ReadyCellProps = {
  player: Player;
};

const ReadyCell = ({
  player: { userId, isReady },
}: ReadyCellProps): JSX.Element => {
  const isCurrentUser = userId === getUserId();

  if (isCurrentUser) {
    return (
      <Checkbox
        checked={isReady}
        onChange={({ target: { checked } }) => {
          setReady({ isReady: checked });
        }}
      />
    );
  }

  if (isReady) {
    return <CheckOutlined css={{ color: 'green' }} />;
  }

  return <CloseOutlined css={{ color: 'red' }} />;
};
export default ReadyCell;
