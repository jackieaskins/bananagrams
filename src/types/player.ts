import { Board } from "./board";
import { Hand } from "./hand";

export enum PlayerStatus {
  NOT_READY = "NOT_READY",
  SPECTATING = "SPECTATING",
  READY = "READY",
}

export type Player = {
  userId: string;
  username: string;
  status: PlayerStatus;
  isTopBanana: boolean;
  isAdmin: boolean;
  gamesWon: number;
  hand: Hand;
  board: Board;
};
