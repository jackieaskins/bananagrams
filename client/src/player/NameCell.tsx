import { CrownOutlined, KeyOutlined } from '@ant-design/icons';
import { Space } from 'antd';

import { Player } from '../players/types';

type NameCellProps = {
  player: Player;
};

const NameCell = ({
  player: { isAdmin, isTopBanana, username },
}: NameCellProps): JSX.Element => (
  <Space>
    <span>{username}</span>
    {isAdmin && (
      <span>
        <KeyOutlined />
      </span>
    )}
    {isTopBanana && (
      <span>
        <CrownOutlined />
      </span>
    )}
  </Space>
);

export default NameCell;
