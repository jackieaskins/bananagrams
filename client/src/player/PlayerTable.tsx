import { Table, Typography } from 'antd';
import { ColumnType } from 'antd/lib/table';

import { useCurrentPlayer, useGamePlayers } from '../game/stateHooks';
import ActionsCell from './ActionsCell';
import NameCell from './NameCell';
import ReadyCell from './ReadyCell';
import { Player } from './types';

const COLUMN_ALIGN = 'center';
const MAX_PLAYERS = 8;

const { Text } = Typography;

const actionsColumn: ColumnType<Player> = {
  align: COLUMN_ALIGN,
  key: 'actions',
  render: function ActionsRender(_: any, player: Player) {
    return <ActionsCell player={player} />;
  },
};

const PlayerTable = (): JSX.Element => {
  const players = useGamePlayers();
  const isAdmin = !!useCurrentPlayer()?.isAdmin;

  return (
    <Table
      bordered
      columns={[
        {
          title: 'Players',
          children: [
            {
              align: COLUMN_ALIGN,
              key: 'ready',
              title: 'Ready?',
              render: function ReadyRender(_, player) {
                return <ReadyCell player={player} />;
              },
            },
            {
              align: COLUMN_ALIGN,
              key: 'player',
              title: 'Player',
              render: function PlayerRender(_, player) {
                return <NameCell player={player} />;
              },
            },
            {
              align: COLUMN_ALIGN,
              dataIndex: 'gamesWon',
              title: 'Games won',
            },
            ...(isAdmin && players.length > 1 ? [actionsColumn] : []),
          ],
        },
      ]}
      dataSource={players}
      footer={() => (
        <div css={{ fontSize: '12px', textAlign: 'center' }}>
          <Text type="secondary">
            The game will start when all players are ready
            <br />
            Current player count: {players.length} / {MAX_PLAYERS}
          </Text>
        </div>
      )}
      pagination={false}
      rowKey="userId"
      size="middle"
    />
  );
};

export default PlayerTable;
