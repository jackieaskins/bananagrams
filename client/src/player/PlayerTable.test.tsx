import { shallow } from 'enzyme';

import { playerFixture } from '../fixtures/player';
import PlayerTable from './PlayerTable';

const mockPlayer = playerFixture({ userId: '1' });
const mockPlayers = [mockPlayer];
jest.mock('../game/stateHooks', () => ({
  useGamePlayers: () => mockPlayers,
}));

describe('<PlayerTable />', () => {
  const renderComponent = (propOverrides = {}) =>
    shallow(<PlayerTable {...propOverrides} />);

  test('renders table', () => {
    expect(renderComponent()).toMatchSnapshot();
  });

  test('renders footer with game info', () => {
    expect(renderComponent().props().footer()).toMatchSnapshot();
  });

  describe('column renders', () => {
    const READY_COLUMN = 0;
    const NAME_COLUMN = 1;
    const ACTIONS_COLUMN = 3;

    const getColumn = (columnIndex: number) =>
      renderComponent()
        .props()
        .columns[0].children[columnIndex].render(null, mockPlayer);

    test('ready column', () => {
      expect(getColumn(READY_COLUMN)).toMatchSnapshot();
    });

    test('name column', () => {
      expect(getColumn(NAME_COLUMN)).toMatchSnapshot();
    });

    test('actions column', () => {
      expect(getColumn(ACTIONS_COLUMN)).toMatchSnapshot();
    });
  });
});
