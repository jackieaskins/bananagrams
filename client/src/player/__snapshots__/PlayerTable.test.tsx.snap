// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayerTable /> column renders actions column does not render if not admin 1`] = `undefined`;

exports[`<PlayerTable /> column renders actions column renders if admin and more than 1 player 1`] = `
<ActionsCell
  player={
    Object {
      "board": Array [
        Array [
          null,
        ],
      ],
      "gamesWon": 0,
      "hand": Array [],
      "isAdmin": false,
      "isReady": false,
      "isTopBanana": false,
      "userId": "1",
      "username": "username",
    }
  }
/>
`;

exports[`<PlayerTable /> column renders name column 1`] = `
<NameCell
  player={
    Object {
      "board": Array [
        Array [
          null,
        ],
      ],
      "gamesWon": 0,
      "hand": Array [],
      "isAdmin": false,
      "isReady": false,
      "isTopBanana": false,
      "userId": "1",
      "username": "username",
    }
  }
/>
`;

exports[`<PlayerTable /> column renders ready column 1`] = `
<ReadyCell
  player={
    Object {
      "board": Array [
        Array [
          null,
        ],
      ],
      "gamesWon": 0,
      "hand": Array [],
      "isAdmin": false,
      "isReady": false,
      "isTopBanana": false,
      "userId": "1",
      "username": "username",
    }
  }
/>
`;

exports[`<PlayerTable /> renders footer with game info 1`] = `
<div
  css="unknown styles"
>
  <Text
    type="secondary"
  >
    The game will start when all players are ready
    <br />
    Current player count: 
    2
     / 
    8
  </Text>
</div>
`;

exports[`<PlayerTable /> renders table 1`] = `
<Table
  bordered={true}
  columns={
    Array [
      Object {
        "children": Array [
          Object {
            "align": "center",
            "key": "ready",
            "render": [Function],
            "title": "Ready?",
          },
          Object {
            "align": "center",
            "key": "player",
            "render": [Function],
            "title": "Player",
          },
          Object {
            "align": "center",
            "dataIndex": "gamesWon",
            "title": "Games won",
          },
        ],
        "title": "Players",
      },
    ]
  }
  dataSource={
    Array [
      Object {
        "board": Array [
          Array [
            null,
          ],
        ],
        "gamesWon": 0,
        "hand": Array [],
        "isAdmin": false,
        "isReady": false,
        "isTopBanana": false,
        "userId": "1",
        "username": "username",
      },
      Object {
        "board": Array [
          Array [
            null,
          ],
        ],
        "gamesWon": 0,
        "hand": Array [],
        "isAdmin": false,
        "isReady": false,
        "isTopBanana": false,
        "userId": "2",
        "username": "username",
      },
    ]
  }
  footer={[Function]}
  pagination={false}
  rowKey="key"
  size="middle"
/>
`;
